mirrord/layer/src/socket/ops.rs

```diff
@@ -1594,26 +1594,56 @@ pub(super) fn getifaddrs() -> HookResult<*mut libc::ifaddrs> {
         Err(io::Error::from_raw_os_error(result))?;
     }
 
-    // Head of the ipv6 interface addresses list.
-    // Should be freed with `libc::freeifaddrs` before returning.
-    let mut ipv6_head: *mut libc::ifaddrs = std::ptr::null_mut();
+    // Count entries for new_list_start malloc
+    let mut entry_count = 0;
+    let mut count_head: *mut libc::ifaddrs = original_head;
+    unsafe {
+        while let Some(ifaddr) = count_head.as_mut() {
+            entry_count += 1;
+            count_head = ifaddr.ifa_next;
+        }
+    }
+
+    // Allocate new list so we can safely free the original list later
+    let new_list_start: *mut libc::ifaddrs = unsafe {
+        libc::malloc((mem::size_of::<libc::ifaddrs>() as libc::size_t) * entry_count)
+            as *mut libc::ifaddrs
+    };
+    // Address to place next new address
+    let mut next_new: *mut libc::ifaddrs = new_list_start;
     // Currently inspected element of the original interface addresses list.
     let mut inspected: *mut libc::ifaddrs = original_head;
-    // Previously inspected element of the original interface addresses list.
-    let mut previous: *mut libc::ifaddrs = std::ptr::null_mut();
+    // Previously element that was inserted into new list.
+    let mut previous_new_entry: *mut libc::ifaddrs = std::ptr::null_mut();
 
     // Safety: we only dereference pointers received from libc. They should be nulls or point to
     // initialized memory.
     unsafe {
         while let Some(ifaddr) = inspected.as_mut() {
             let address = SockaddrStorage::from_raw(ifaddr.ifa_addr, None);
+
             match address.as_ref().and_then(SockaddrStorage::as_sockaddr_in6) {
                 // If not ipv6, advance to the next interface address in the original list.
                 // Move both `previous` and `inspected`.
                 None => {
-                    previous = inspected;
-                    inspected = ifaddr.ifa_next;
-                    continue;
+                    // Append the address to the new list by copying ifaddr to the list head, and
+                    // setting next of previous then moving head
+                    copy_nonoverlapping::<libc::ifaddrs>(inspected, next_new, 1);
+                    match previous_new_entry.as_mut() {
+                        Some(prev_addr) => {
+                            prev_addr.ifa_next = next_new;
+                        }
+                        None => {
+                            // inspected is head of list
+                        }
+                    }
+
+                    previous_new_entry = next_new;
+                    // clippy lint fails on this line but the pointer arithmetic is fine
+                    #[allow(clippy::size_of_in_element_count)]
+                    {
+                        next_new = next_new.wrapping_add(mem::size_of::<libc::ifaddrs>());
```

Comment:
```suggestion
                    {
                        next_new = next_new.add(mem::size_of::<libc::ifaddrs>());
```

Can't we just use `ptr::add` here instead? We control the allocations, offsets, we should be able to satisfy `add`.